Deserialize for dynamically sized struct generates invalid code

```rust
extern crate serde;

#[macro_use]
extern crate serde_derive;

#[derive(Deserialize)]
struct S {
    last: [u8],
}
```

```
error[E0277]: the trait bound `[u8]: std::marker::Sized` is not satisfied in `S`
 --> src/main.rs:6:10
  |
6 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ within `S`, the trait `std::marker::Sized` is not implemented for `[u8]`
  |
  = note: `[u8]` does not have a constant size known at compile-time
  = note: required because it appears within the type `S`
  = note: required by `serde::Deserialize`
```

--- a/serde_derive/src/de.rs
+++ b/serde_derive/src/de.rs
@@ -15,15 +15,16 @@ use syn::{self, Ident, Index, Member};
 use bound;
 use fragment::{Expr, Fragment, Match, Stmts};
 use internals::ast::{Container, Data, Field, Style, Variant};
-use internals::{self, attr};
+use internals::{attr, Ctxt};
 use pretend;
 use try;
 
 use std::collections::BTreeSet;
 
 pub fn expand_derive_deserialize(input: &syn::DeriveInput) -> Result<Tokens, String> {
-    let ctxt = internals::Ctxt::new();
+    let ctxt = Ctxt::new();
     let cont = Container::from_ast(&ctxt, input);
+    precondition(&ctxt, &cont);
     try!(ctxt.check());
 
     let ident = cont.ident;
@@ -78,6 +79,16 @@ pub fn expand_derive_deserialize(input: &syn::DeriveInput) -> Result<Tokens, Str
         };
     };
     Ok(generated)
+}
+
+fn precondition(cx: &Ctxt, cont: &Container) {
+    if let Data::Struct(_, ref fields) = cont.data {
+        if let Some(last) = fields.last() {
+            if let syn::Type::Slice(_) = *last.ty {
+                cx.error("cannot deserialize a dynamically sized struct");
+            }
+        }
+    }
 }
 
 struct Parameters {

--- a/test_suite/tests/compile-fail/precondition/deserialize_dst.rs
+++ b/test_suite/tests/compile-fail/precondition/deserialize_dst.rs
@@ -1,0 +1,16 @@
+// Copyright 2018 Serde Developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[macro_use]
+extern crate serde_derive;
+
+#[derive(Deserialize)] //~ ERROR: proc-macro derive panicked
+struct S {
+    string: String,
+    slice: [u8], //~^^^ HELP: cannot deserialize a dynamically sized struct
+}

